plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java

```diff
@@ -445,4 +452,57 @@ public void renameMaterializedView(ConnectorSession session, SchemaTableName sou
     {
         throw new TrinoException(NOT_SUPPORTED, "renameMaterializedView is not supported for Iceberg Glue catalogs");
     }
+
+    @Override
+    public Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(tableName, "tableName is null");
+        Optional<String> targetCatalogName = getHiveCatalogName(session);
+        if (targetCatalogName.isEmpty()) {
+            return Optional.empty();
+        }
+        if (isHiveSystemSchema(tableName.getSchemaName())) {
+            return Optional.empty();
+        }
+
+        // we need to chop off any "$partitions" and similar suffixes from table name while querying the metastore for the Table object
+        int metadataMarkerIndex = tableName.getTableName().lastIndexOf('$');
+        SchemaTableName tableNameBase = (metadataMarkerIndex == -1) ? tableName : schemaTableName(
+                tableName.getSchemaName(),
+                tableName.getTableName().substring(0, metadataMarkerIndex));
+
+        Optional<com.amazonaws.services.glue.model.Table> table = getTable(tableNameBase.getSchemaName(), tableNameBase.getTableName());
+
+        if (table.isEmpty()) {
+            return Optional.empty();
+        }
+        if (isIcebergTable(table.get())) {
+            // After redirecting, use the original table name, with "$partitions" and similar suffixes
+            return targetCatalogName.map(catalog -> new CatalogSchemaTableName(catalog, tableName));
+        }
+        return Optional.empty();
+    }
+
+    private static boolean isIcebergTable(com.amazonaws.services.glue.model.Table table)
+    {
+        return ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(table.getParameters().get(TABLE_TYPE_PROP));
+    }
+
+    private Optional<com.amazonaws.services.glue.model.Table> getTable(String databaseName, String tableName)
+    {
+        try {
+            GetTableResult result = stats.getGetTable().call(() ->
+                    glueClient.getTable(new GetTableRequest()
+                            .withDatabaseName(databaseName)
+                            .withName(tableName)));
+            return Optional.of(result.getTable());
+        }
+        catch (EntityNotFoundException e) {
+            return Optional.empty();
+        }
+        catch (AmazonServiceException e) {
+            throw new TrinoException(HIVE_METASTORE_ERROR, e);
+        }
+    }
```

Comment:
Why not use  the existing `private Optional<com.amazonaws.services.glue.model.Table> getTable(SchemaTableName schemaTableName)`?

Reply:
It wasn't yet in the code at the time when I wrote the initial draft of the PR. Making use now of the shared method.